### ISSUE TITLE
SB2209 / SB2240 Won't go into DFU Mode

### ISSUE THREAD

**Issue by alexrizk1117** [bigtreetech/EBB#74]
I'm trying to install an sb2209 board using U2C wired into a pi. I'm trying to get the board into DFU mode but haven't had any success. All I see when running "lsusb" on my pi is:

- OpenMoko, Inc. stm32f445xx

- OpenMoko, Inc. Geschwister Schneider CAN adapter

Also, I'm trying to use the Boot + Reset method, but that doesn't seem to work. Is the SB2209 supposed to have a solid green and red light?

Also, when having the U2C board hooked up, should it be running power through the 24V PS, or should that only be powered through the pi in usb C?

**Comment by Nooni99** [comment]
Have you found a solution @alexrizk1117 ? I am having the same issue except I see " Raspberry Pi RP2 Boot" when I try to go to DFU mode. Despite this, I am unable to enter dfu mode.

**Comment by ga73** [comment]
@Nooni99 i'm having the same issue...have you worked this out ?

**Comment by Nooni99** [comment]
@ga73 I have not found the exact solution and I just gave up on it because I wanted to get my printer working and just wired everything directly to my motherboard without canboot. However, I will do some experimenting once I have time. The next thing I am going to try which might work is using stm32 cube programmer to rewrite the sb2209. That might work and I suggest you try it. Let me know the results.

**Comment by ga73** [comment]
Okay I got everything working last night and now up and running... I connected both the utc and rp2209 to 24v and ground instead of 5v usb power... The only other thing I did different was to get into dfu on the rp2209 I held down the boot button and then pressed the reset for maybe 3 second.. Let go of reset and then boot button... My can wires went to the screw terminals on the utc... Green was can low and yellow can high... I hope this helps someone

**Comment by ga73** [comment]
Also make sure you remove the usb 5v jumper BEFORE connecting to 24v and that you have the 120r jumpers on both the utc and rp2209 board otherwise they won't communicate with each other

**Comment by ga73** [comment]
Also this link was much better than just using the BTT guide alone...

https://github.com/Esoterical/voron_canbus

**Comment by ga73** [comment]
Your welcome, I hope it helps, I'm not expert but feel free to leave another message here if you get stuck and I'll see what I did differently.. Good luck

**Comment by ParabolicTech** [comment]
I am having the same issue. I hold the boot button and press the reset and it doesnt go into DFU mode. The device ID shows up with no description. This happens weather or not I try to enable DFU. I can plug it into my windows machine and it WILL bring up the bootloader folder if I do the button sequence. But for some odd reason connected to my Raspberry PI it doesnt function into DFU mode. Anyone else with this issue? Is there a way I can flash it in Windows?

**Comment by Nooni99** [comment]
@ParabolicTech have you tried the solutions from above?

**Comment by ParabolicTech** [comment]
> @ParabolicTech have you tried the solutions from above?

Yes, Thats why I posted. Ive been at this for hours in multiple places and cant figure out why I cant enable DFU. Diffrent known working cables - different (5v) power supplies, in fact I have my raspPI disconnected from my octopus board JUST to eliminate some of the other addresses showing up.

**Comment by Nooni99** [comment]
@ParabolicTech hmmm. That's interesting. I'm not sure what's wrong. Those boards are quite difficult to get running anyways and I hope bigtreetech does another version that's easier to get working. In the mean time, I hope someone posts a solution that works for you. Best of luck!

**Comment by ga73** [comment]
You mention you connected to 5v..have you tried connecting to 24v? This seemed to make a difference for me, also make sure your usb cable is a data and not just a usb power cable and finally do you have the r120 jumpers on both your canbus board AND your UTC (assuming this is what your using)?

I connected mine up fully and then flashed it, saying that if you just can't get it working try stm32cube software on windows as a last resort

**Comment by xPakrikx** [comment]
same problem ... nothing helps. i see only Raspberry Pi RP2 Boot under lsusb.

**Comment by LeighT616** [comment]
Hi i've just hit this issue... the board will not go into DFU mode.  I've tried various USB C cables and zero dice!

is there any resolution on this, i'm looking into the stm32cub software via windows but surely there has to be a better solution

**Comment by ParabolicTech** [comment]
So after a lot of trials and tribulations I found out it DOES go into DFU mode by holding the reset button and clicking the boot button once. HOWEVER, the feedback that's listed online isn't correct. The board im using has the RP2040 chip on it so it wont list the description when it goes into DFU mode as listed on various instructions. The description is just blank. I was able to continue with the process and flashed the board, got the green light and the flashing red light and the rest of the listed setup worked fine.

**Comment by rafal-molotkiewicz** [comment]
I have the same issue. My board only shows in two modes:

`Bus 001 Device 058: ID 1d50:614e OpenMoko, Inc. rp2040`
or
`Bus 001 Device 057: ID 2e8a:0003 Raspberry Pi RP2 Boot`

Neither one shows when I try `sudo dfu-util -l`

I tried holding boot and short pressing reset, long pressing reset, double and triple click reset.
I tried above swapping buttons.

I tried all this with 5V power, 24V power, and 24V power with connected CAN data lines.

Nothing made any difference. How can I proceed from here?

**Comment by Nooni99** [comment]
Yeah, I'm still having the same issue with getting it in DFU mode. I have tried everything.

**Comment by wes4560** [comment]
I have my sb2209 rb2040 hooked directly to the octopus using the rj connection I have unplugged everything used the 5v jumper and also tried the 24 v supply and I can't get it to go in DFU mode the light just stays green and I can't get it to show up at all in the ssh with the pi

**Comment by Nooni99** [comment]
@wes4560 I found the solution. Since the chip is an RP2040, it will use boot mode. Your device should appear as "Raspberry Pi RP2 Boot" which is effectively the same as DFU mode. You can download firmware to it with no problem. Best of luck!

**Comment by mprenter** [comment]
Same issue as everyone above, having a brand new board and unable to get it into DFU mode.  Can get the device to say "RP2 Boot" which some people say is good enough, but how do I install the firmware without using the dfu-util?

**Comment by Nooni99** [comment]
@mprenter please don't make the misunderstanding I did. "RP2 Boot" is the "DFU mode" of this board. Since the board has a RP chip, it uses "RP2 Boot" instead. So follow all the steps you should follow and once you have "RP2 Boot" you're ready to move onto the next steps. You should be able to install the firmware the same you would any other chip. Let me know if you have more questions.

**Comment by looxonline** [comment]
> Same issue as everyone above, having a brand new board and unable to get it into DFU mode. Can get the device to say "RP2 Boot" which some people say is good enough, but how do I install the firmware without using the dfu-util?

This is not an ST chip. It is an RP chip. You can't use dfu-util with it, you need to use the programming methods mentioned in the manual.

**Comment by Destra-Slaybot** [comment]
> Have you found a solution @alexrizk1117 ? I am having the same issue except I see " Raspberry Pi RP2 Boot" when I try to go to DFU mode. Despite this, I am unable to enter dfu mode.

was this ever solved, i can't get past this one part

**Comment by bigtreetech** [comment]
> > Have you found a solution @alexrizk1117 ? I am having the same issue except I see " Raspberry Pi RP2 Boot" when I try to go to DFU mode. Despite this, I am unable to enter dfu mode.
> 
> was this ever solved, i can't get past this one part

If you see this, then it is in DFU. It is an RP2040. The manual has the info you need.

**Comment by lunchbox2883** [comment]
Same situation here... Not getting the DFU mode but i am getting the PI RP2 boot. but getting the same error as @Charleslotto above when pushing through with the flash

**Comment by Jax1er** [comment]
I have never had this issue before but am now.  Not sure what I am doing wrong.  I can put the 2209 into what looks like DFU mode with a red flashing light but lsusb doesn't change either before or after.
Bus 008 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 007 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 001 Device 011: ID 1d50:614e OpenMoko, Inc. stm32g0b1xx
Bus 001 Device 010: ID 1d50:614e OpenMoko, Inc. stm32g0b1xx
Bus 001 Device 006: ID 0bda:5411 Realtek Semiconductor Corp. RTS5411 Hub
Bus 001 Device 004: ID 0bda:5411 Realtek Semiconductor Corp. RTS5411 Hub
Bus 001 Device 009: ID 1d50:606f OpenMoko, Inc. Geschwister Schneider CAN adapter
Bus 001 Device 002: ID 1a40:0101 Terminus Technology Inc. Hub
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 004 Device 001: ID 1d6b:0001 Linux Foundation 1.1 root hub
Bus 003 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 006 Device 001: ID 1d6b:0001 Linux Foundation 1.1 root hub
Bus 005 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub

**Comment by dahpuffy** [comment]
Hey everyone, I recently went through this exact same headache with my own RP2040 board and finally figured out the missing link.
The confusion here is happening because RP2040 chips do not use dfu-util. They don't work like the STM32 chips you might be used to, so running make flash FLASH_DEVICE=... will always fail (as @Charleslotto saw with the Invalid DFU suffix signature error).

When your board shows up as Raspberry Pi RP2 Boot via lsusb, it is actually fully ready to be flashed. In this mode, the RP2040 acts exactly like a standard USB thumb drive. You don't use a flash tool; you literally just copy and paste the firmware file onto it!

Here is the exact process:
1. Compile the Firmware
Ensure your make menuconfig is set to Raspberry Pi RP2040, then run make. Notice that for RP2040s, Klipper generates a klipper.uf2 file instead of a .bin file.

2. Enter Boot Mode
Hold the BOOT button while plugging in the USB cable(the process on my Eddy duo)(or hold BOOT, press RST, release BOOT). Verify lsusb shows Raspberry Pi RP2 Boot.

3. Copy the File
Because it's acting like a USB drive, Linux usually auto-mounts it. Look for a drive named RPI-RP2 (usually under /media/pi/RPI-RP2/ or /media/biqu/RPI-RP2/). 
Just copy the file over:
cp ~/klipper/out/klipper.uf2 /media/<your-username>/RPI-RP2/

4. That's It!
The moment the .uf2 file finishes transferring, the board will automatically disconnect the USB drive and reboot into Klipper.

(For Windows users like @ParabolicTech: It's the same process. When you plug it in while holding BOOT, a folder will pop up on your desktop. Just drag and drop the klipper.uf2 file into that folder and you're done).

Hope this saves someone the hours of troubleshooting it cost me!

**Comment by Jax1er** [comment]
I assume this is for RB2240 and not SB2209/2240.  this is not working for me

**Comment by dahpuffy** [comment]
Hey @Jax1er, sorry my bad. I looked at your lsusb output again and realized why my fix isn't working for you.

​BTT actually makes two versions of your board: one with an RP2040 chip, and one with an STM32 chip. Your lsusb shows stm32g0b1xx, meaning you have the STM32 version. The .uf2 drag-and-drop method I posted is only for the RP2040.

​The good news is that seeing "OpenMoko, Inc." usually means your board already has Katapult or Klipper installed, so I think you should be able to skip the DFU entirely and just flash your firmware over CAN or USB using the Katapult script:
python3 ~/katapult/scripts/flashtool.py ...

I had some problems with the BOOT and RESET buttons, so if you are struggling to get you board into DFU mode with the RESET button, try powering the board off completely. Press and hold the BOOT button, turn the power back on, and then release BOOT. This worked for me at least.

I don't have your exact board, but my EBB36 uses the same STM32 chip. I'm definitely no expert, but I literally just spent hours back-and-forth with Gemini with a similar issue when setting up CAN between my Kraken, EBB36 and Eddy Duo. 
I don't know your exact hardware and wiring to give you the exact command, but I hope this points you in the right direction!

**Comment by Jax1er** [comment]
No worries!  Thanks for your help and time!  Yes I have the SB2209/2240 CAN board with he 2240 driver.  Comms are going to my main board USB via I2C.  The step I was missing was that I had to unplug the CAN cable and then plug in the USB-C cable directly to my pi.  I saw the instructions and (been a long time) forgot the board has a USB-C port on it.  It’s in the back side of the board and hidden by my umbilical connection holder.  Strange though, I don’t remember having to do that the last time because I probably just updated through Katapult as you said.  I figured it out this morning after sleeping on it.  Another funny thing that happened was when I transferred over my printer config file , my extruder SB 2240 SPI set up section was missing.  Everything in Klipper (mainsail) said it was fine. I did the printer tune for my bed scanner/touch and started the first print.  As the printer started printing, I noticed there was no filament coming out of the nozzle.  I canceled the print and did a manual motor test and noticed the extruder wasn’t moving.   After quite a bit of digging and head scratching, I noticed the extruder driver section was missing from the config!  Go figure.
I’m up and running now, so hope this helps someone down the road.  Also I am not sure if Klipper should’ve given me some kind of notice saying that I was missing the driver section for extruder.  🤷‍♂️ 
Thanks again
